### PR TITLE
feat: enhance verifiers and expose util functions

### PR DESCRIPTION
**Pull request by Taowyoo**
- Change verifiers to be able to ignore expired cert.
- Change verifiers to be able to have a custom verify callback.
- Make some util functions public.

### CODE REVIEW

File: rustls-mbedpki-provider/src/client_cert_verifier.rs [hunk 1]
```diff
@@ -23,6 +25,8 @@ use crate::{
 pub struct MbedTlsClientCertVerifier {
     trusted_cas: mbedtls::alloc::List<mbedtls::x509::Certificate>,
     root_subjects: Vec<rustls::DistinguishedName>,
+    verify_callback: Option<Arc<dyn mbedtls::x509::VerifyCallback + 'static>>,
+    ignore_expired: bool,
```
Comment (s-arash) [inline rustls-mbedpki-provider/src/client_cert_verifier.rs:29]: Can we add a type, something like `struct CertActiveCheck {pub ignore_expired: bool, pub ignore_not_active_yet: bool}`?

File: rustls-mbedpki-provider/src/client_cert_verifier.rs [hunk 2]
```diff
@@ -62,6 +66,43 @@ impl MbedTlsClientCertVerifier {
...
     }
+
+    /// Sets the optional verification callback function for the certificate verification process.
+    ///
+    /// The verification callback allows you to customize how the certificate verification is performed.
+    ///
+    /// # Arguments
+    ///
+    /// * `callback` - A trait object implementing the `mbedtls::x509::VerifyCallback` trait, wrapped in an `Arc`.
+    pub fn set_verify_callback(&mut self, callback: Arc<dyn mbedtls::x509::VerifyCallback + 'static>) {
```
Comment (s-arash) [inline rustls-mbedpki-provider/src/client_cert_verifier.rs:77]: Can this take an `Option`? I'm thinking of trying to reset `verify_callback`

File: rustls-mbedpki-provider/src/client_cert_verifier.rs [hunk 3]
```diff
@@ -92,11 +133,26 @@ impl ClientCertVerifier for MbedTlsClientCertVerifier {
...
+                    None,
+                    Some(&mut error_msg),
+                    move |cert: &mbedtls::x509::Certificate, depth: i32, flags: &mut mbedtls::x509::VerifyError| {
+                        callback(cert, depth, flags)
+                    },
+                )
+                .map_err(|e| mbedtls_err_into_rustls_err_with_error_msg(e, &error_msg))?;
+            }
+            None => mbedtls::x509::Certificate::verify(&chain, &self.trusted_cas, None, Some(&mut error_msg))
+                .map_err(|e| mbedtls_err_into_rustls_err_with_error_msg(e, &error_msg))?,
```
Comment (s-arash) [inline rustls-mbedpki-provider/src/client_cert_verifier.rs:154]: Can this line (and line 151) be moved outside of `match {...}`?

File: rustls-mbedpki-provider/src/client_cert_verifier.rs (hunk 2 above)
Comment (Taowyoo) [reply rustls-mbedpki-provider/src/client_cert_verifier.rs:77]: sure

File: rustls-mbedpki-provider/src/client_cert_verifier.rs (hunk 1 above)
Comment (Taowyoo) [reply rustls-mbedpki-provider/src/client_cert_verifier.rs:29]: make sense

File: rustls-mbedpki-provider/src/client_cert_verifier.rs (hunk 3 above)
Comment (Taowyoo) [reply rustls-mbedpki-provider/src/client_cert_verifier.rs:144]: done!

File: rustls-mbedpki-provider/src/client_cert_verifier.rs (hunk 2 above)
Comment (Taowyoo) [reply rustls-mbedpki-provider/src/client_cert_verifier.rs]: done!

File: rustls-mbedpki-provider/src/client_cert_verifier.rs (hunk 1 above)
Comment (Taowyoo) [reply rustls-mbedpki-provider/src/client_cert_verifier.rs]: done!

File: rustls-mbedpki-provider/src/lib.rs [hunk 4]
```diff
@@ -20,6 +20,21 @@ pub mod server_cert_verifier;
...
+/// A config about whether to check certificate validity period
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct CertActiveCheck {
+    /// If accept expired certificate, default to false
+    pub ignore_expired: bool,
+    /// If accept not active certificate, default to false
+    pub ignore_not_active_yet: bool,
+}
+
+impl Default for CertActiveCheck {
```
Comment (s-arash) [inline rustls-mbedpki-provider/src/lib.rs:32]: This could be derived.

File: rustls-mbedpki-provider/src/lib.rs [hunk 5]
```diff
@@ -20,6 +20,21 @@ pub mod server_cert_verifier;
 pub use client_cert_verifier::MbedTlsClientCertVerifier;
 pub use server_cert_verifier::MbedTlsServerCertVerifier;
 
+/// A config about whether to check certificate validity period
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct CertActiveCheck {
+    /// If accept expired certificate, default to false
```
Comment (s-arash) [inline rustls-mbedpki-provider/src/lib.rs:26]: Maybe could be "Accept expired certificates"?

File: rustls-mbedpki-provider/src/lib.rs [hunk 6]
```diff
@@ -20,6 +20,21 @@ pub mod server_cert_verifier;
...
+    pub ignore_expired: bool,
+    /// If accept not active certificate, default to false
```
Comment (s-arash) [inline rustls-mbedpki-provider/src/lib.rs:28]: Could be "Accept certificates that are not yet active"

**Review by s-arash — APPROVED** [review]

File: rustls-mbedpki-provider/src/lib.rs (hunk 4 above)
Comment (Taowyoo) [reply rustls-mbedpki-provider/src/lib.rs]: done

File: rustls-mbedpki-provider/src/lib.rs (hunk 5 above)
Comment (Taowyoo) [reply rustls-mbedpki-provider/src/lib.rs]: done

File: rustls-mbedpki-provider/src/lib.rs (hunk 6 above)
Comment (Taowyoo) [reply rustls-mbedpki-provider/src/lib.rs]: done

**Pull request merged**